### PR TITLE
feat(smime): sign outgoing emails

### LINKED ISSUE #7671: Sign messages with S/MIME
### Is your feature request related to a problem? Please describe.

As a user with an S/MIME certificate I want to be able to sign my outgoing messages so senders can verify the authenticity of my messages.

### Describe the solution you'd like

The app should offer an option in the UI that enables creation of an S/MIME signature. When a message is sent the app will then sign it with the user's certificate.

- [x] Requires https://github.com/nextcloud/mail/issues/7667 - specifically the private key

The trust model is to protect messages outside the boundaries of an organization. Therefore the server and its admins are trusted. Public and private keys may be known in plain text on the server-side.

### Describe alternatives you've considered

N/a

### Additional context

This is part of https://github.com/nextcloud/mail/issues/7085

### PR DESCRIPTION
**Pull request by st3iny**
Fix #7671

## Testing

1. Generate self signed certificates via https://github.com/nextcloud/mail/blob/main/doc/developer.md#generate-self-signed-smime-ca-and-user-certificates.
2. Import the CA certificate to make the verification work: occ security:certificates:import /path/to/ca.crt

Happy testing!

## TODO

- [x] Add a certificate select for the account/primary alias (can only be set for aliases atm)
- [x] Rename `*SMime*` to `*Smime*`

## Design feedback

* [x] If there’s no private key, do not show the entry
* [x] If there is one, show it in the action menu above "Request read receipt"
* [x] Simple checkbox with label: "Sign message"
* [ ] Should be checked by default and remembered per-account -> Followup
* [ ] When we add encryption, directly below would be another checkbox labeled "Encrypt message" -> Separate PR for encryption
* [ ] When we move actions like Attachments and Formatting out of the menu, signing and encrypting should still stay in there as they are technical -> Followup

### CODE REVIEW

**Review by ChristophWurst — COMMENTED** [review]
Exceptionally clean diff.

I will give this a test as well

File: lib/Model/EnrichedSMimeCertificate.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ */
+
+namespace OCA\Mail\Model;
+
+use JsonSerializable;
+use OCA\Mail\Db\SMimeCertificate;
+use ReturnTypeWillChange;
+
+class EnrichedSMimeCertificate implements JsonSerializable {
```
[the remark this answers is not in the document]
Comment (st3iny) [reply lib/Model/EnrichedSMimeCertificate.php:32]: On second thought, Smime makes more sense.

**Comment by st3iny** [comment]
*Potential bad UX:* Scheduling a message and deleting the associated certificate results in failing to send the message. The message has to be opened and resent again in the frontend (with a new cert id or signing has to be disabled). Nothing is broken, the user just never gets notified about the reason of the failure and how to fix it.

I prefer to keep it this way to prevent sending unsigned messages when a user explicitly requested a signature.

Server will log warnings:
```
OCA\Mail\Exception\ServiceException: Could not send message: Certificate does not exist:
Did expect one result but found none when executing:
query "SELECT * FROM `*PREFIX*mail_smime_certificates` WHERE (`id` = :dcValue1) AND (`user_id` = :dcValue2)"; 
```

File: appinfo/routes.php [hunk 2]
```diff
@@ -94,6 +95,11 @@
 			'url' => '/api/accounts/{id}/signature',
 			'verb' => 'PUT'
 		],
+		[
+			'name' => 'accounts#updateSmimeCertificate',
+			'url' => '/api/accounts/{id}/smimeCertificate',
```
Comment (ChristophWurst) [inline appinfo/routes.php:100]: nit: smime-certificate? camel case in an URL feels a bit strange

File: src/components/AliasForm.vue [hunk 3]
```diff
@@ -92,89 +129,165 @@ export default {
...
+			// Only show certificates that are at least valid until tomorrow
+			const now = (new Date().getTime() / 1000) + 3600 * 24
 
-			await this.$store.dispatch('updateAlias', {
-				account: this.account,
-				aliasId: this.alias.id,
-				alias: this.changeAlias,
-				name: this.changeName,
-			})
+			return this.smimeCertificates
```
Comment (ChristophWurst) [inline src/components/AliasForm.vue:174]: I have the feeling this always return an empty array with the self-signed cert. At least I'm not able to test currently. There is no cert selector in my alias settings.

**Comment by ChristophWurst** [comment]
> **Potential bad UX:** Scheduling a message and deleting the associated certificate

Catch associated outbox message before a cert is deleted and abort the operation?

File: src/components/AliasForm.vue (hunk 3 above)
Comment (kesselb) [reply src/components/AliasForm.vue:174]: Same for me :confused: 

```
openssl_x509_checkpurpose($certificate, X509_PURPOSE_SMIME_SIGN, $caBundle),
openssl_x509_checkpurpose($certificate, X509_PURPOSE_SMIME_ENCRYPT, $caBundle),
```

SIGN and ENCRYPT are both false.

- Used gen-smime-cert.sh to generate CA and certificates.
- Did not import the CA. 
- Uploaded both certificate to mail.

Comment (kesselb) [reply src/components/AliasForm.vue:174]: I'm able to select the certificate when importing the CA via occ.

Signing a message also works :partying_face: 

The signature is shown in Thunderbird as untrusted. After importing the CA it's shown as valid.

**Comment by ChristophWurst** [comment]
> 1) OCA\Mail\Tests\Unit\Controller\MessagesControllerTest::testGetHtmlBody
> PHPUnit\Framework\MockObject\UnknownTypeException: Class or interface "OCA\Mail\Service\SMimeService" does not exist

File: src/components/AliasForm.vue (hunk 3 above)
Comment (st3iny) [reply src/components/AliasForm.vue:174]: Interesting. So the checkpurposes method also verifies the local certificate chain.

I updated the instructions above.

**Comment by st3iny** [comment]
> > **Potential bad UX:** Scheduling a message and deleting the associated certificate
> 
> Catch associated outbox message before a cert is deleted and abort the operation?

I did some more testing and our backend logic is already smart enough. It sets the failed column and won't attempt to resend it.


## Query to find pending messages
```php
$qb->expr()->orX(
	$qb->expr()->isNull('failed'),
	$qb->expr()->eq('failed', $qb->createNamedParameter(false, IQueryBuilder::PARAM_BOOL), IQueryBuilder::PARAM_BOOL),
)
```

## Screenshot from the outbox

grafik

File: appinfo/routes.php (hunk 2 above)
Comment (st3iny) [reply appinfo/routes.php]: Done

**Comment by ChristophWurst** [comment]
Nit: add ext-openssl to composer.json to make tools happy

> 'ext-openssl' is missing in composer.json 

It is required for Nextcloud and even shipped by default with PHP8.0 https://docs.nextcloud.com/server/stable/admin_manual/installation/source_installation.html

**Review by ChristophWurst — APPROVED** [review]
Works!

**Pull request merged**